File: 9-Stack/programs/ruby/program5.rb
# Program to convert  Infix to postfix Conversion

class Node
  attr_accessor :data, :next

  def initialize(data)
    @data = data
    @next = nil
  end
end

class Stack
  def initialize
    @top = nil
  end

  def push(x)
    t = Node.new(x)
    if t.nil?
      puts "Stack is full"
    else
      t.next = @top
      @top = t
    end
  end

  def pop
    x = nil
    if @top.nil?
      puts "Stack is Empty"
    else
      t = @top
      @top = @top.next
      x = t.data
    end
    x
  end

  def display
    p = @top
    while p
      print "#{p.data} "
      p = p.next
    end
    puts
  end

  def top
    @top ? @top.data : nil
  end
end

def precedence(x)
  case x
  when '+', '-'
    1
  when '*', '/'
    2
  else
    0
  end
end

def is_operand(x)
  !['+', '-', '*', '/'].include?(x)
end

def infix_to_postfix(infix)
  stack = Stack.new
  postfix = ""
  stack.push('#')
  infix.each_char do |char|
    if is_operand(char)
      postfix += char
    else
      while precedence(char) <= precedence(stack.top)
        postfix += stack.pop
      end
      stack.push(char)
    end
  end
  while stack.top != '#'
    postfix += stack.pop
  end
  postfix
end

infix = "a+b*c-d/e"
puts infix_to_postfix(infix)

File: 9-Stack/programs/cpp/program5.cpp
// Program to convert  Infix to postfix Conversion
#include<bits/stdc++.h>
using namespace std;

struct Node{
  char data;
  struct Node *next;
}*top = NULL;

void push(char x){
  struct Node *t;
  t = new Node;
  if(t == NULL){
    cout << "Stack is full" << endl;
  }else{
    t->data = x;
    t->next = top;
    top = t;
  }
}

char pop(){
  char x = -1;
  if(top == NULL){
    cout << "Stack is Empty" << endl;
  }else{
    struct Node *t = top;
    top = top->next;
    x = t->data;
    delete t;
  }
  return x;
}

void Display(){
  struct Node *p = top;
  while(p != NULL){
    cout << p->data << " ";
    p = p->next;
  }
  cout << endl;
}

int pre(char x){
  if(x == '+' || x == '-'){
    return 1;
  }else if(x == '*' || x == '/'){
    return 2;
  }
  return 0;
}

int isOperand(char x){
  if(x == '+' || x == '-' || x == '*' || x == '/'){
    return 0;
  }else{
    return 1;
  }
}


string InToPost(string infix){
  string postfix = "";
  int i = 0;
  while(i < infix.length()){
    if(isOperand(infix[i])){
      postfix += infix[i];
      i++;
    }else{
      if(pre(infix[i]) > pre(top->data)){
        push(infix[i++]);
      }else{
        postfix += pop();
      }
    }
  }
  while(top != NULL){
    postfix += pop();
  }
  return postfix;
}

int main(){
  string infix = "a+b*c-d/e";
  push('#');
  cout << InToPost(infix) << endl;
  return 0;
}

// Output
// abc*+de/-#